### PR TITLE
Cache IVF-PQ and select-warpsort kernel launch parameters to reduce latency

### PR DESCRIPTION
**Pull request by achirkin**
This PR aims at reducing the latency in IVF-PQ and related functions, especially with small work sizes and in the "throughput" benchmark mode.

 - Add kernel config caching to ivf_pq::search::compute_similarity kernel
 - Add kernel config caching to select::warpsort
 - Fix the memory_resource usage in `matrix::select_k`: make sure all temporary allocations use raft's workspace memory resource.

### CODE REVIEW

**Comment by achirkin** [comment]
I've benchmarked prims::SelectK, prims:KNN/ivf_pq (with n_queries set to 1), and ann-bench ivf-pq.

Performance changes in the prims benchmarks are very minimal; most cases a the same withing +-2-3% time, a few cases became faster by 5-8%.

The most benefit comes in ann-bench run in "throughput" mode. At 4 threads, it's 20%, at 32 threads it's 60% faster.
IVF-PQ speed up in the throughput mode

**Review by cjnolet — APPROVED** [review]
LGTM.

**Review by tfeher — APPROVED** [review]
Thanks for this PR Artem, it looks good to me. I have one question, but that need not block merging this.

File: cpp/include/raft/neighbors/detail/ivf_pq_search.cuh [hunk 1]
```diff
@@ -408,6 +415,46 @@ constexpr inline auto expected_probe_coresidency(uint32_t n_clusters,
...
+         a.preferred_shmem_carveout == b.preferred_shmem_carveout && a.pq_bits == b.pq_bits &&
+         a.pq_dim == b.pq_dim && a.precomp_data_count == b.precomp_data_count &&
+         a.n_queries == b.n_queries && a.n_probes == b.n_probes && a.topk == b.topk;
+}
+
+struct search_kernel_key_hash {
+  inline auto operator()(const search_kernel_key& x) const noexcept -> std::size_t
+  {
+    return (size_t{x.manage_local_topk} << 63) +
+           size_t{x.topk} * size_t{x.n_probes} * size_t{x.n_queries} +
```
[the remark this answers is not in the document]
Comment (achirkin) [reply cpp/include/raft/neighbors/detail/ivf_pq_search.cuh:442]: Good point! But does raft depend on boost already? There seem to be no standard library equivalent for this.
Also, in the current implementation of LRU cache, we don't use hash values at all (we changed the implementation from a hashmap to a vector of key-value tuples).

**Pull request merged**